src/detect-ftp-mode.c

```diff
@@ -0,0 +1,166 @@
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectFtpModeMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags, void *state,
+        void *txv, const Signature *s, const SigMatchCtx *m)
+{
+    FTPTransaction *tx = (FTPTransaction *)txv;
+    if (tx->command_descriptor.command_code == FTP_COMMAND_UNKNOWN) {
+        return 0;
+    }
+    if (!tx->dyn_port) {
+        return 0;
+    }
+
+    const DetectFtpModeData *ftpmoded = (const DetectFtpModeData *)m;
+    return 1 ? ftpmoded->active == tx->active : 0;
+}
+
+/**
+ * \brief this function will free memory associated with DetectFtpModeData
+ *
+ * \param ptr pointer to DetectFtpModeData
+ */
+static void DetectFtpModeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+/**
+ * \brief This function is used to parse ftp.mode options passed via ftp.mode keyword
+ *
+ * \param str Pointer to the user provided ftp.mode options
+ *
+ * \retval  pointer to DetectFtpModeData on success
+ * \retval NULL on failure
+ */
+static DetectFtpModeData *DetectFtpModeParse(const char *optstr)
+{
+    DetectFtpModeData *ftpmoded = SCCalloc(1, sizeof(DetectFtpModeData));
+    if (unlikely(ftpmoded == NULL))
+        return NULL;
+
+    char *token = SCParseToken(optstr, strlen(optstr));
+    if (!token) {
+        SCLogError("unable to extract token value from \"%s\"", optstr);
+        goto error;
+    } else if (!strcmp(token, "active")) {
+        ftpmoded->active = true;
+    } else if (!strcmp(token, "passive")) {
+        ftpmoded->active = false;
+    } else {
+        SCLogError("Invalid command value");
+        goto error;
+    }
+
+    SCFreeToken(token);
+    return ftpmoded;
+
+error:
+    if (token) {
+        SCFreeToken(token);
+    }
+    if (ftpmoded)
+        SCFree(ftpmoded);
+    return NULL;
+}
+
+static int DetectFtpModeSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectSignatureSetAppProto(s, ALPROTO_FTP) != 0)
+        return -1;
+
+    DetectFtpModeData *ftpmoded = DetectFtpModeParse(str);
+    if (ftpmoded == NULL)
+        return -1;
+
+    if (SigMatchAppendSMToList(de_ctx, s, DETECT_FTP_MODE, (SigMatchCtx *)ftpmoded,
+                g_ftp_mode_buffer_id) == NULL) {
+        DetectFtpModeFree(de_ctx, ftpmoded);
+        return -1;
+    }
+
+    return 0;
+}
+
+void DetectFtpModeRegister(void)
+{
+    /* ftp.mode sticky buffer */
+    sigmatch_table[DETECT_FTP_MODE].name = KEYWORD_NAME;
+    sigmatch_table[DETECT_FTP_MODE].desc = "sticky buffer to match on the FTP mode buffer";
+    sigmatch_table[DETECT_FTP_MODE].url = "/rules/" KEYWORD_DOC;
+    sigmatch_table[DETECT_FTP_MODE].Setup = DetectFtpModeSetup;
+    sigmatch_table[DETECT_FTP_MODE].AppLayerTxMatch = DetectFtpModeMatch;
+    sigmatch_table[DETECT_FTP_MODE].Free = DetectFtpModeFree;
+
+    DetectAppLayerInspectEngineRegister(
```

Comment:
mode is based on the command, not on the reply, right?

Reply:
Yes -- where the command is PORT, EPRT, PASV, EPSV